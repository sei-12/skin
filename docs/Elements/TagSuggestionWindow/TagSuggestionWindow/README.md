[**skin**](../../../README.md)

***

[skin](../../../modules.md) / Elements/TagSuggestionWindow/TagSuggestionWindow

# Elements/TagSuggestionWindow/TagSuggestionWindow

## Index

### Namespaces

- [TagSuggestionWindow](namespaces/TagSuggestionWindow/README.md)
