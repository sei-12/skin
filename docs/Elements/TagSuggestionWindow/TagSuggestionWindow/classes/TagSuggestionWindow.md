[**skin**](../../../../README.md)

***

[skin](../../../../modules.md) / [Elements/TagSuggestionWindow/TagSuggestionWindow](../README.md) / TagSuggestionWindow

# Class: TagSuggestionWindow

## Constructors

### new TagSuggestionWindow()

> **new TagSuggestionWindow**(`settings`?): [`TagSuggestionWindow`](TagSuggestionWindow.md)

#### Parameters

##### settings?

[`TagSuggestionWindowSettings`](../interfaces/TagSuggestionWindowSettings.md)

#### Returns

[`TagSuggestionWindow`](TagSuggestionWindow.md)

#### Defined in

[Elements/TagSuggestionWindow/TagSuggestionWindow.ts:39](https://github.com/sei-12/skin/blob/71b214be76b363d9a6ca32cbed9404c037be92a0/src/Elements/TagSuggestionWindow/TagSuggestionWindow.ts#L39)

## Properties

### root

> **root**: `HTMLElement`

#### Defined in

[Elements/TagSuggestionWindow/TagSuggestionWindow.ts:51](https://github.com/sei-12/skin/blob/71b214be76b363d9a6ca32cbed9404c037be92a0/src/Elements/TagSuggestionWindow/TagSuggestionWindow.ts#L51)

## Methods

### focusDown()

> **focusDown**(): `void`

#### Returns

`void`

#### Defined in

[Elements/TagSuggestionWindow/TagSuggestionWindow.ts:66](https://github.com/sei-12/skin/blob/71b214be76b363d9a6ca32cbed9404c037be92a0/src/Elements/TagSuggestionWindow/TagSuggestionWindow.ts#L66)

***

### focusUp()

> **focusUp**(): `void`

#### Returns

`void`

#### Defined in

[Elements/TagSuggestionWindow/TagSuggestionWindow.ts:63](https://github.com/sei-12/skin/blob/71b214be76b363d9a6ca32cbed9404c037be92a0/src/Elements/TagSuggestionWindow/TagSuggestionWindow.ts#L63)

***

### getNumItems()

> **getNumItems**(): `number`

#### Returns

`number`

#### Defined in

[Elements/TagSuggestionWindow/TagSuggestionWindow.ts:60](https://github.com/sei-12/skin/blob/71b214be76b363d9a6ca32cbed9404c037be92a0/src/Elements/TagSuggestionWindow/TagSuggestionWindow.ts#L60)

***

### updateItems()

> **updateItems**(`itemDatas`): `void`

#### Parameters

##### itemDatas

[`TagSuggestionWindowItemData`](../interfaces/TagSuggestionWindowItemData.md)[]

#### Returns

`void`

#### Defined in

[Elements/TagSuggestionWindow/TagSuggestionWindow.ts:53](https://github.com/sei-12/skin/blob/71b214be76b363d9a6ca32cbed9404c037be92a0/src/Elements/TagSuggestionWindow/TagSuggestionWindow.ts#L53)
