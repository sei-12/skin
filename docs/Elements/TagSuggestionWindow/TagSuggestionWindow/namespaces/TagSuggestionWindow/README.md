[**skin**](../../../../../README.md)

***

[skin](../../../../../modules.md) / [Elements/TagSuggestionWindow/TagSuggestionWindow](../../README.md) / TagSuggestionWindow

# TagSuggestionWindow

## Index

### Classes

- [Element](classes/Element.md)

### Interfaces

- [Setting](interfaces/Setting.md)
- [TagFinder](interfaces/TagFinder.md)
