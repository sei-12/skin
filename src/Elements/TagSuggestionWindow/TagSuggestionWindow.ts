import { Assert } from "../../common/Assert"
import { h } from "../../common/dom"
import { scroll_to_focus_elm } from "../../common/scroll"
import {  I_CommandEmmiter } from "../../lib/CommandEmmiter"
import { CommandEmiterListener } from "../../lib/EmiterCore"
import styles from "./style.module.css"


/**
 * static関数にするべきかもしれないが
 * namespaceの外から参照して欲しくないから
 * functionで書く
 */
function joinTextBlocks(blocks: TextBlock[]){
    let text = ""
    blocks.forEach( b => { text += b.text })
    return text
}

function generateTextBlockElm(text: TextBlock, settings: TagSuggestionWindow.Setting){
    const elm = h("span")
    elm.root.innerText = text.text
    if ( text.match ){
        elm.root.style.color = settings.matchTextColor
    }else{
        elm.root.style.color = settings.unmatchTextColor
    }
    return elm
}

function generateItemElm(item: ItemData,settings: TagSuggestionWindow.Setting){
    const elm = h(`div.${styles.itemelm}`,[
        
    ])
    
    let textBlockElms = item.textBlocks().map( t => generateTextBlockElm(t,settings) )
    
    textBlockElms.forEach( e => {
        elm.root.appendChild(e.root)
    })
    
    return elm
}

class FocusIndex {

    private value: number | null = null
    private targetAryLength: number | null = null
    
    moveFocus(to: "up" | "down"){
        if ( this.targetAryLength === 0 ){
            return
        }
        Assert.isNotNull(this.value)
        Assert.isNotNull(this.targetAryLength)

        this.value += to === "up" ? -1 : 1

        if (to === "up" ? this.value < 0 : this.value >= this.targetAryLength){
            this.value = to === "up" ? this.targetAryLength -1 : 0
        }
    }

    /**
     * 要素の型は問わない
     */
    updateTargetAry(ary: Array<unknown>){
        this.targetAryLength = ary.length

        if ( ary.length == 0 ){
            this.value = null
            return
        }

        if ( this.value === null ){
            this.value = 0
            return
        }
        
        if ( ary.length <= this.value ){
            this.value = ary.length - 1
            return
        }
    }
    
    getOrThrow(): number {
        Assert.isNotNull(this.value)
        return this.value
    }
    
    get(): number | null {
        return this.value
    }
}

class TextBlock {
    constructor(
        public readonly text: string,
        public readonly match: boolean
    ){ }
}

class ItemData {
    constructor(
        private predicate: string, 
        private tagText: string
    ){ }

    textBlocks(): TextBlock[] {
        let splited = this.tagText.split(this.predicate)
        if ( splited.length === 1 ){
            splited.unshift(this.predicate)
        }else if ( splited.length > 1 ){
            let newSplited = []
            newSplited.push(splited.shift()!)
            newSplited.push(this.predicate)
            newSplited.push(splited.join(this.predicate))
            splited = newSplited
        }else{
            throw new Error("?")
        }
        return splited.map( text => new TextBlock(text, text === this.predicate))
    }
}

class TagFinderWraper {
    constructor(
        private readonly finder: TagSuggestionWindow.TagFinder
    ){}
    
    async find(predicate: string){
        let result = await this.finder.find(predicate)
        let itemDatas = result.map( res => new ItemData(predicate, res) )
        return itemDatas
    }
}

class DefaultSettings implements TagSuggestionWindow.Setting {
    matchTextColor: string = "white"
    unmatchTextColor: string = "gray"
    width: number = 200;
    maxHeightPx: number = 350;
    backgroundColor: string = "rgb(20,20,20)"
    focusBackgroundColor: string = "rgb(40,40,40)"
}


//----------------------------------------------------------------------------------------------------//
//                                                                                                    //
//                                               PUBLIC                                               //
//                                                                                                    //
//----------------------------------------------------------------------------------------------------//

export namespace TagSuggestionWindow {

    export interface Setting {
        readonly width: number
        readonly maxHeightPx: number
        readonly backgroundColor: string
        readonly focusBackgroundColor: string
        readonly matchTextColor: string
        readonly unmatchTextColor: string
    }

    /**
     * DBのラッパーがこのインターフェースを満たしたクラスのインスタンスを生成する予定
     */
    export interface TagFinder {
        find(predicate: string): Promise<string[]>
    }
    
    export class Element {
        private focusIndex = new FocusIndex()
        private settings: TagSuggestionWindow.Setting
        private tagFinderWraper: TagFinderWraper
        private elm = h(`div.${styles.root}`,[
            
        ])
        
        private items: {
            data: ItemData
            elm:ReturnType<typeof generateItemElm>
        }[] = []


        private clearFocusFromHtml(){
            this.items[this.focusIndex.getOrThrow()].elm.root.style.backgroundColor = ""
        }
        private focusHtml(){
            this.items[this.focusIndex.getOrThrow()].elm.root.style.backgroundColor = this.settings.focusBackgroundColor
            scroll_to_focus_elm(
                this.items[this.focusIndex.getOrThrow()].elm.root,
                this.elm.root
            )
        }
        private updateItems(itemDatas: ItemData[]){
            this.items = itemDatas.map( data => {
                return {
                    elm: generateItemElm(data,this.settings),
                    data
                }
            })
            

            this.elm.root.innerHTML = ""

            this.focusIndex.updateTargetAry(this.items)

            if ( this.items.length === 0 ){
                this.hide()
                return
            }else{
                this.show()
            }

            this.items.forEach( item => {
                this.elm.root.appendChild(item.elm.root)
            })
            
            this.focusHtml()
        }
        
        private hide(){
            this.elm.root.style.display = "none"
        }
        private show(){
            this.elm.root.style.display = "block"
        }
        
        private listener: CommandEmiterListener

        //
        // public
        //
        constructor(
            tagFinder: TagFinder,
            commandEmmiter: I_CommandEmmiter,
            settings?: TagSuggestionWindow.Setting
        ){
            
            this.listener = new CommandEmiterListener(
                ["tagSuggestionWindow.focusDown",() => { this.moveFocus("down") }],
                ["tagSuggestionWindow.focusUp",() => { this.moveFocus("up") }],
            )

            commandEmmiter.addWeakRefListener(this.listener)

            this.tagFinderWraper = new TagFinderWraper(tagFinder)

            if ( settings ){
                this.settings = settings
            }else{
                this.settings = new DefaultSettings()
            }
            

            this.elm.root.style.width = `${this.settings.width}px`
            this.elm.root.style.maxHeight = `${this.settings.maxHeightPx}px`
            this.elm.root.style.backgroundColor = this.settings.backgroundColor
            

            this.hide()
        }

        root: HTMLElement = this.elm.root


        async update(predicate: string){
            let itemDatas = await this.tagFinderWraper.find(predicate)
            this.updateItems(itemDatas)
        }

        private moveFocus(to: "up" | "down"){
            if ( this.focusIndex.get() === null ){
                return
            }

            this.clearFocusFromHtml()
            this.focusIndex.moveFocus(to)
            this.focusHtml()
        }

        setWindowPos(topPx:number, leftPx: number){
            this.elm.root.style.left = leftPx + "px"
            this.elm.root.style.top  = topPx + "px"
        }

        getFocused(): string | null{
            let index = this.focusIndex.get()
            if ( index === null ){
                return null
            }
            return joinTextBlocks(this.items[index].data.textBlocks())
        }
    }
}