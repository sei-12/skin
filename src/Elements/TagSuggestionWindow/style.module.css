.root {
    position: absolute;
    overflow-y: scroll;
    overflow-x: hidden;
    box-sizing: border-box;
    font-family: Hack Nerd Font;
    font-weight: bold;
}

.root::-webkit-scrollbar {
    display: none;
}

.itemelm {
    padding: 2px;
    text-wrap: nowrap;
}