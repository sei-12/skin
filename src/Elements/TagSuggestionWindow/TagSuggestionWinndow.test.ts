import { expect, it } from "vitest";
import { TagSuggestionWindow } from "./TagSuggestionWindow";

class DecoyItemData implements TagSuggestionWindow.ItemData {
    textBlocks(): TagSuggestionWindow.TextBlock[] {
        return [
            new TagSuggestionWindow.TextBlock("hel",true),
            new TagSuggestionWindow.TextBlock("l",true),
            new TagSuggestionWindow.TextBlock("o",true),
        ]
    }    
}

it("TagSuggestionWindow:whitebox",() => {
    const elm = new TagSuggestionWindow.Element()
    
    let numDatas = 100
    let datas = Array(numDatas).fill( new DecoyItemData() )
    elm.updateItems(datas)
    
    expect(elm.root.style.maxHeight).toBe("150px")
    expect(elm.root.style.width).toBe("200px")
    expect(((elm as any).elm.root as HTMLElement).childNodes.length).toBe(numDatas)
})

it("TagSuggestionWindow:blackbox",() => {
    const elm = new TagSuggestionWindow.Element()
    
    let numDatas = 100
    let datas = Array(numDatas).fill( new DecoyItemData() )
    elm.updateItems(datas)
    
    expect(elm.getFocused()).toBe("hello")
})