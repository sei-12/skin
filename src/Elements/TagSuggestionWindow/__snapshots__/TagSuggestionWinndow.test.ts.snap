// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TagSuggestionWindow 1`] = `null`;

exports[`TagSuggestionWindow 2`] = `null`;

exports[`TagSuggestionWindow 3`] = `"gcp"`;

exports[`TagSuggestionWindow 4`] = `null`;

exports[`TagSuggestionWindow 5`] = `null`;

exports[`TagSuggestionWindow 6`] = `null`;

exports[`TagSuggestionWindow 7`] = `"postgresql"`;

exports[`TagSuggestionWindow 8`] = `null`;

exports[`TagSuggestionWindow 9`] = `null`;

exports[`TagSuggestionWindow 10`] = `null`;

exports[`TagSuggestionWindow 11`] = `"github"`;

exports[`TagSuggestionWindow 12`] = `null`;

exports[`TagSuggestionWindow 13`] = `"optimizations"`;

exports[`TagSuggestionWindow 14`] = `"matplotlib"`;

exports[`TagSuggestionWindow 15`] = `"javascript"`;

exports[`TagSuggestionWindow 16`] = `"ci/cd"`;
